lbrynet/core/HTTPBlobDownloader.py

```diff
@@ -21,65 +22,76 @@ def __init__(self, blob_manager, blob_hashes=None, servers=None, client=None):
     def download_blob(self, blob):
-        try:
-            yield self._download_blob(blob)
-            self.failures = 0
-        except Exception as exception:
-            self.failures += 1
-            log.exception('Mirror failed downloading')
-            if self.failures >= self.max_failures:
-                self.stop()
-                self.failures = 0
-
-    @defer.inlineCallbacks
-    def _download_blob(self, blob):
-        if not blob.get_is_verified() and not blob.is_downloading() and 'mirror' not in blob.writers:
-            response = yield self.client.get(url_for(choice(self.servers), blob.blob_hash))
-            if response.code != 200:
-                log.debug('[Mirror] Missing a blob: %s', blob.blob_hash)
-                if blob.blob_hash in self.blob_hashes:
-                    self.blob_hashes.remove(blob.blob_hash)
-                defer.returnValue(blob.blob_hash)
-            log.debug('[Mirror] Download started: %s', blob.blob_hash)
-            blob.set_length(response.length)
+        for _ in range(self.max_failures):
             writer, finished_deferred = blob.open_for_writing('mirror')
+            self.writers.append(writer)
             try:
-                yield self.client.collect(response, writer.write)
-                log.info('Mirror completed download for %s', blob.blob_hash)
-            except Exception as e:
-                writer.close(e)
-            yield finished_deferred
-            defer.returnValue(blob.blob_hash)
+                downloaded = yield self._write_blob(writer, blob)
+                if downloaded:
+                    yield finished_deferred  # yield for verification errors, so we log them
+                    if blob.verified:
+                        log.info('Mirror completed download for %s', blob.blob_hash)
+                break
+            except (IOError, Exception) as e:
+                if isinstance(e, DownloadCanceledError) or 'closed file' in str(e):
+                    # some other downloader finished first or it was simply cancelled
+                    log.info("Mirror download cancelled: %s", blob.blob_hash)
+                    break
+                else:
+                    log.exception('Mirror failed downloading')
+            finally:
+                finished_deferred.addBoth(lambda _: None)  # suppress echoed errors
+                if 'mirror' in blob.writers:
+                    writer.close()
+                self.writers.remove(writer)
+
+
+    @defer.inlineCallbacks
+    def _write_blob(self, writer, blob):
+        response = yield self.client.get(url_for(choice(self.servers), blob.blob_hash))
+        if response.code != 200:
+            log.debug('[Mirror] Missing a blob: %s', blob.blob_hash)
+            if blob.blob_hash in self.blob_hashes:
+                self.blob_hashes.remove(blob.blob_hash)
+            defer.returnValue(False)
+
+        log.debug('[Mirror] Download started: %s', blob.blob_hash)
```

Comment:
the logging formatter will include the file and lineno, so `[Mirror]` isn't necessary

Reply:
:man_facepalming: yes, of course! fixing it!